### PR TITLE
Weak-ref-ify MetaConverter and FakeTensorConverter

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80545
* __->__ #80544

Make `MetaConverter` and `FakeTensorConverter` hold weak references to their memoized tensors, and also have `MetaConverter` hold weak reference to Tensor storage. Otherwise it can be tricky for users to make sure all existing FakeTensors or FakeTensorModes are deleted which otherwise will leak memory. 

I ran into https://github.com/pytorch/pytorch/issues/7733 which I was able to get around with the following (see comment from code):

```
# torch.Tensors cannot be used as a key in a dictionary
# because they define a custom __eq__ function which when used
# to resolve hash collisions will throw when comparing tensors:
# "RuntimeError: bool value of Tensor with more than one value is ambiguous."
# To avoid that, we use an object which will hold a Tensor and use
# its id for both hashing and equality.
# In order to use this as a weak key reference, we cannot
# simply use weakref.WeakKeyDictionary because the newly constructed
# WeakTensorRefKey only use would be a dictionary so it would have no strong
# references.
# To get around this issue, we can use it as a normal key, and then set
# `weakref.finalize` to delete the key when its contained tensor dies.
``` 

While for the tensor memo we can set a `weakref.finalize` callback that will remove the corresponding `WeakTensorRefKey` from the tensor memo, at the point that this callback is invoked the tensor storage is not yet deallocated.. See comment from code:

```
# [expired-storages]
# NB: even though the tensor has died,
# the deallocation of its storage can take longer,
# even when the storage has no other uses/views.
# In this case, the StorageWeakRef object will be kept alive
# longer than it needs to be, however the storage itself
# will be deallocated. We retain the possibly dead storages
# and periodically check if any of them are expired and
# can be freed.
```

partial fix for https://github.com/pytorch/torchdynamo/issues/468

### CODE REVIEW

File: torch/_subclasses/meta_utils.py [hunk 1]
```diff
@@ -9,36 +11,133 @@ def safe_is_leaf(t):
...
+
+class WeakTensorRefKey(object):
+    def __init__(self, ten):
+        self.ten = weakref.ref(ten)
+        # store id since as soon as ten is deallocated
+        # the old id will no longer be recoverable, and
+        # we need to be able to remove the WeakTensorRefKey
+        # from the dictionary by hashing it to the same
+        # value it had when ten was alive
+        self.id = id(self.ten())
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py:35]: This seems like it is vulnerable to ABA problem where you may incorrectly conclude this key is equal to another Python object that got reallocated to the same id. I need to read the rest of the code, but it sounds like there is a correctness constraint where WeakTensorRefKey must be promptly disposed when ten dies.

File: torch/_subclasses/meta_utils.py [hunk 2]
```diff
@@ -9,36 +11,133 @@ def safe_is_leaf(t):
...
+            else:
+                # [expired-storages]
+                # NB: even though the tensor has died,
+                # the deallocation of its storage can take longer,
+                # even when the storage has no other uses/views.
+                # In this case, the StorageWeakRef object will be kept alive
+                # longer than it needs to be, however the storage itself
+                # will be deallocated. We retain the possibly dead storages
+                # and periodically check if any of them are expired and
+                # can be freed.
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py:114]: Wouldn't a more conventional design be to attach finalizer to the storage, and maintain storage memo removal separately?

**Review by ezyang — APPROVED** [review]
The implementation is a little circuitous but it looks alright to ship

File: torch/_subclasses/meta_utils.py (hunk 2 above)
Comment (eellison) [reply torch/_subclasses/meta_utils.py:114]: As of now at least, there's no binding of Storage in python that would be deallocated when it's corresponding tensors are deallocated, as far as I'm aware of. that would be nice though..

**Pull request closed without merging**